File: Beelina.LIB/Models/Product.cs
using System.ComponentModel.DataAnnotations.Schema;
using Beelina.LIB.Interfaces;

namespace Beelina.LIB.Models
{
    public class Product
        : Entity, IUserActionTracker
    {
        public string Name { get; set; }
        public string Code { get; set; }
        public string Description { get; set; }
        public int StockQuantity { get; set; }
        public float PricePerUnit { get; set; }
        public float CostPrice { get; set; }
        public bool IsTransferable { get; set; }
        public int NumberOfUnits { get; set; }
        public DateTime ValidFrom { get; set; }
        public DateTime? ValidTo { get; set; }
        public bool Parent { get; set; }
        public int? ProductParentGroupId { get; set; }
        public virtual Product ProductParentGroup { get; set; }

        [NotMapped]
        public bool IsLinkedToSalesAgent { get; set; }
        [NotMapped]
        public double SearchResultPercentage { get; set; }

        public double Price
        {
            get
            {
                return Math.Round(PricePerUnit, 2);
            }
        }

        public double Cost
        {
            get
            {
                return Math.Round(CostPrice, 2);
            }
        }

        [NotMapped]
        public bool IsCurrentlyActive
        {
            get
            {
                DateTime now = DateTime.UtcNow.Date;
                return ValidFrom <= now && (!ValidTo.HasValue || ValidTo > now);
            }
        }

        public int ProductUnitId { get; set; }
        public ProductUnit ProductUnit { get; set; }

        public int SupplierId { get; set; }
        public Supplier Supplier { get; set; }

        public int? DeletedById { get; set; }
        public virtual UserAccount DeletedBy { get; set; }
        public int? UpdatedById { get; set; }
        public virtual UserAccount UpdatedBy { get; set; }
        public int? CreatedById { get; set; }
        public virtual UserAccount CreatedBy { get; set; }
        public int? DeactivatedById { get; set; }
        public virtual UserAccount DeactivatedBy { get; set; }
    }
}

File: Beelina.APP/src/app/_models/product.ts
import { NumberFormatter } from '../_helpers/formatters/number-formatter.helper';
import { IModelNode } from '../_interfaces/imodel-node';
import { Entity } from './entity.model';
import { ProductUnit } from './product-unit';

export class Product extends Entity implements IModelNode {
  public name: string;
  public code: string;
  public description: string;
  public supplierId: number;
  public stockQuantity: number;
  public pricePerUnit: number;
  public costPrice: number;
  public price: number;
  public cost: number;
  public productUnit: ProductUnit;
  public deductedStock: number;
  public withdrawalSlipNo: string;
  public plateNo: string;
  public isTransferable: boolean;
  public numberOfUnits: number;
  public isLinkedToSalesAgent: boolean;
  public validFrom?: Date;
  public validTo?: Date;
  public parent?: boolean;
  public productParentGroupId?: number;

  get priceFormatted(): string {
    return NumberFormatter.formatCurrency(this.price);
  }

  get costFormatted(): string {
    return NumberFormatter.formatCurrency(this.cost);
  }

  get nameWithUnit() {
    return `${this.productUnit.name} - ${this.name}`;
  }

  get isCurrentlyActive(): boolean {
    if (!this.validFrom) {
      return false;
    }

    const now = new Date();
    const validFromDate = new Date(this.validFrom);
    const validToDate = this.validTo ? new Date(this.validTo) : null;

    // Set time to start/end of day for proper comparison
    now.setHours(0, 0, 0, 0);
    validFromDate.setHours(0, 0, 0, 0);

    if (validToDate) {
      validToDate.setHours(23, 59, 59, 999);
    }

    // Product is active if:
    // 1. validFrom date has passed (validFrom <= now)
    // 2. validTo is null OR validTo date hasn't passed yet (validTo >= now)
    return validFromDate <= now && (!validToDate || validToDate >= now);
  }

  constructor() {
    super();
    this.productUnit = new ProductUnit();
  }
}
